bittensor/dendrite.py

```diff
@@ -250,50 +299,110 @@ def __init__(self, neuron: bittensor_pb2.Neuron, config, keypair):
+                    Single torch tensor to be sent to the remote neuron endpoint.
+
+                mode (:obj:`bittensor_pb2.Modality` of shape :obj:`(1)`, `required`):
+                    Bittensor forward modality type. Enum in [TEXT, IMAGE, TENSOR]
+
+            Returns:
+                output (:obj:`Tuple[torch.FloatTensor, torch.LongTensor]`, `required`):
+                    Result tuple from the forward call.
+
+        """
+        # ---- On Backoff: We dont make an RPC and return zeros instead ----  
         if self.config.dendrite.do_backoff and self.backoff >= 1:
-            self.backoff -= 1
-            logger.trace('Still backing off from endpoint {}', self.endpoint)
-            return nill_response_for(inputs), torch.tensor(bittensor_pb2.ReturnCode.Backoff)
-    
-        try:
-            outputs, code = _RemoteModuleCall.apply(self, DUMMY, inputs, mode)
+            outputs = nill_response_for(inputs)
+            code = torch.tensor(bittensor_pb2.ReturnCode.Backoff)
 
-            # Update backoff on non-success.
-            if code.item() != bittensor_pb2.ReturnCode.Success and self.config.dendrite.do_backoff:
-                self.backoff = self.next_backoff
-                self.next_backoff = min(self.config.dendrite.max_backoff, self.next_backoff * 2)
 
-            return outputs, code
+        # ---- On Not-backoff: We make the Forward RPC ---- 
+        else:
+            try:
+                # Make and time the query.
+                start_time = time.time()
+                outputs, code = _RemoteModuleCall.apply(self, DUMMY, inputs, mode)
+                elapsed_time = time.time() - start_time
 
-        except Exception as e:
-            error_msg = 'Uncaught error in forward call with error {}'.format( e )
-            logger.error(error_msg)
-            return nill_response_for(inputs), torch.tensor(bittensor_pb2.ReturnCode.UnknownException)
+            # ---- On unknown failure: we return zeros and unknown code ---- 
+            except Exception as e:
+                logger.error('Uncaught error in forward call with error {}'.format( e ))
+                outputs = nill_response_for(inputs)
+                code = torch.tensor(bittensor_pb2.ReturnCode.UnknownException)
+
+        # ---- Update stats ---- 
+        self.stats.codes[code.item()] += 1
+        if code.item() == bittensor_pb2.ReturnCode.Success:
+            # We only need to record elapsed times / bytes on success.
+            self.stats.total_success_time += elapsed_time
+            self.stats.avg_success_time = self.stats.total_success_time / self.stats.codes[bittensor_pb2.ReturnCode.Success]
+            self.stats.success_bytes_out += sys.getsizeof(inputs)
+            self.stats.success_bytes_in += sys.getsizeof(outputs)
+
+        # ---- On Success: set zero backoff and halve the next backoff ---- 
+        if code.item() == bittensor_pb2.ReturnCode.Success:
```

Comment:
Guessing you rebased this off the #114  change? :) same comment applies, not sure why we're checking twice for the same return code.